### ISSUE TITLE
Potential logical error in info_nce_loss implementation

### ISSUE THREAD

**Issue by argaaw** [fchest/M3ANet#4]
Hello, thank you for sharing the code for M3ANet.

While studying the info_nce_loss function in your implementation, I noticed a part that seems to deviate from the standard Contrastive Learning logic and Equation (6) in your paper.

In the current code, pos_logits are summed into a scalar value before being used in the loss calculation:

```python
pos_logits = torch.exp(F.cosine_similarity(eeg_features, img_features, dim=-1) / tau)
pos_logits = torch.sum(pos_logits)  # <--- Summed into a single scalar
```

Mathematically, this makes the loss function behave as follows for the $i$-th sample:

$$ \mathcal{L}'_i = -\log \frac{\sum_{j=1}^{N} \exp(sim(q_j, x_j^+)/\tau)}{\left(\sum_{j=1}^{N} \exp(sim(q_j, x_j^+)/\tau)\right) + \sum_{k \neq i} \exp(sim(q_i, x_k)/\tau)} $$

If `pos_logits` becomes a scalar (the sum of all positive scores in the batch), the numerator in the loss function becomes the "sum of all positive pairs" rather than the "specific positive pair score for the current sample $i$."

This implies that the loss might be minimized by optimizing only a few strong pairs in the batch to increase the global sum, rather than aligning each EEG segment with its corresponding speech segment individually. This seems inconsistent with Eq. (6), which defines the loss element-wise for each sample pair $(q_i, x_i^+)$.

$$ \mathcal{L}_{InfoNCE} = -\log \frac{\exp(sim(q_i, x_i^+)/\tau)}{\sum_{k=1}^{n} \exp(sim(q_i, x_k)/\tau)} $$

To align the code with the paper, pos_logits must remain a vector of size (Batch_Size,) to correspond element-wise with neg_logits.

```python
# Proposed Fix
pos_logits = torch.exp(F.cosine_similarity(eeg_features, img_features, dim=-1) / tau)
```

Could you please clarify if this global summation was intended, or if it is an implementation error?

Thank you for your time.

**Comment by fchest** [comment]
Thank you for your attention to our work! Due to an oversight, we mistakenly uploaded the in-development code intended for subsequent debugging and failed to verify it in a timely manner. This issue has now been fixed.

**Comment by argaaw** [comment]
Thank you for your response. That’s very helpful.
